qos_controller/serenity_controller.cpp

```diff
@@ -22,25 +30,57 @@ class SerenityControllerProcess :
     public Process<SerenityControllerProcess> {
  public:
   SerenityControllerProcess(
-      const lambda::function<Future<ResourceUsage>()>& _usage)
-      : usage(_usage) {}
+      const lambda::function<Future<ResourceUsage>()>& _usage,
+      std::shared_ptr<QoSControllerPipeline> _pipeline)
+    : usage(_usage), pipeline(_pipeline) {}
 
-  Future<list<QoSCorrection>> corrections() {
+  Future<QoSCorrections> corrections() {
     return this->usage()
       .then(defer(self(), &Self::_corrections, lambda::_1));
   }
 
-  Future<list<QoSCorrection>> _corrections(
+  Future<QoSCorrections> _corrections(
       const Future<ResourceUsage>& _resourceUsage) {
-    // TODO(bplotka) Set up the main qos correction pipeline here.
-    std::cout << "Serenity QoS Controller pipeline run." << "\n";
+    QoSCorrections corrections = this->__corrections(_resourceUsage);
+
+    // TODO(bplotka): We don't want to spam slave with empty corrections.
+    // We could make 20 iterations and try to find correction like this:
+    //    this->iterations = 0;
+    //    while(corrections.empty() && this->iterations < 20) {
+    //      std::cout << "xxxx "  << std::endl;
+    //
+    //      os::sleep(Duration::create(60).get());
+    //
+    //      ResourceUsage usage = this->usage().get();
+    //      corrections = this->__corrections(usage);
+    //      this->iterations++;
+    //    }
+    // I would like to hear your opinions about that first.
+
+    return corrections;
+  }
+
+  QoSCorrections __corrections(
+      const Future<ResourceUsage>& _resourceUsage) {
+    Try<QoSCorrections> ret = this->pipeline->run(_resourceUsage.get());
+
+    QoSCorrections corrections;
+
+    if (ret.isError()) {
+      LOG(ERROR) << ret.error();
+      // Return empty corrections.
+    } else {
+      corrections = ret.get();
+    }
 
-    // For now return unsatisfied Future.
-    return Future<list<QoSCorrection>>();
+    return corrections;
   }
 
  private:
   const lambda::function<Future<ResourceUsage>()> usage;
+  std::shared_ptr<QoSControllerPipeline> pipeline;
+  //! Safeguard against infinite loop.
```

Comment:
Sharp eye! I know there is a flag in slave.. qos_controller_interval, however if we set longer duration, it could give us some latency in correction in some situation.